### DISCUSSION TITLE
Create HTMX-enabled components?

### DISCUSSION THREAD

**Discussion by nerdoc** [django-components/django-components#310] [Ideas]
Hello. I was just playing around with the idea of extending django-components with some HTMX functionality, to create "interactive" components.

Call me crazy.

There are already some libs that fill that gap in the net like [Unicorn](https://django-unicorn.com), [Tetra](https://tetraframework.com)(dead?), [Reactor](https://github.com/edelvalle/reactor)(instable/dead?) and others (TurboDjango(dead?), Sockpuppet(dead?) etc). And [djhtmx](https://github.com/edelvalle/djhtmx)(dead?).

What would it take to make an "enhanced" component that 
* is re-rendered after a hx-get/post in-place
* features methods that are "callable" by the client using htmx calls.

Some ideas:
* Component must have a root tag with an id (e.g. hash from class)
*  global urlpattern to "call" component methods, like `__components__/<component_name>/<method>/` that points to
* a view that finds the components from the registry, and calls their requested method, and returns the rendered component  using `hx-swap`
```python
# urls.py
urlpatterns = [path("__components__/<component>/<method>/", component_dispatcher, name="components")]
```
```python
# components/counter.py
class HTMXComponent(Component):
    def get_context_data(self, *args, **kwargs):
        context = super().get_context_data(*args, **kwargs)
        context["component_id"] = f"component_id_{hash(self.__class__)}"
         # or even return a full root tag attr list as string, with id, hx attrs etc.
        return context

class Counter(HTMXComponent):
    template_name = "components/counter.html"
    value: int = 0
    def inc(self, v):
        self.value += v
    def dec(self, v):
        self.value -= v
```
```django
<!-- components/counter.html -->
<div id="{{ component_id }}">
  <button hx-post="{% some-template-tag-helper-here 'inc' %}">inc</button>
    <input type="number" value={{ value }}>
  <button hx-post="{% some-template-tag-helper-here 'inc' %}">dec</button>
</div>
```
The first 2 are simple to implement (already done here as a test), for the 3rd I have no idea, and some things seem to burst my head:
* how to keep track of the "state" of the component, like a counter value?
* how to render the components, reusing as much code as possible: using the do_component template tag?

Thanks for your feedback...

**Comment by rtr1** [comment]
@nerdoc I like the idea. In your example, would values like `Counter.value` correspond to a Django model for persistence, or would it be more transient/temporary, like a value in memory?

In thinking about this, it seems like there are two separate things we are bundling here:

- Component - Bundling HTML, CSS, Javascript
- Data Manipulation - Bundling CRUD and other custom functions and binding to endpoints

If the data manipulation is more model-focused, then maybe the custom class-based view (similar to `CRUDView` below) is an option, or going with the "fat model" approach where these methods (`inc`, `dec`, etc) are methods on the Django model itself (or added via some meta inline class, mixin, etc).

As for creating the views, the library [neapolitan](https://github.com/carltongibson/neapolitan) has functionality where it populates CRUD views, which sounds similar to what you describe. We could have it add any provided functions, instead of just CRUD functions. The point being, `get_urls()` could return a list of endpoints like `path('inc/<int:counter_id>/', ...)`, and `path('dec/<int:counter_id>/', ...)`, and whatever else.

Not that we would use neapolitan, but it serves as an example of this sort of thing.

```python
# urls.py
from neapolitan.views import CRUDView

class BookmarkView(CRUDView):
    model = Bookmark
    fields = ["url", "title", "note"]
    filterset_fields = [
        "favourite",
    ]

urlpatterns = [ ... ] + BookmarkView.get_urls()```

**Comment by EmilStenstrom** [comment]
Sorry for not responding sooner, totally missed this thread.

I really like this idea too, and it’s something I’ve been thinking about for a while.

Some thoughts:
* I think it’s important to **not** make things HTMX-specific, and instead build a general system for re-rendering components. Having examples how to do things with HTMX is fine, as long as it’s possible to switch out if you want to use react or something.
* I think it would be worthwhile to supply some static glue JS that talks to the server and re-renders so the user doesn’t have to pick a framework and all that (similar to Phoenix LiveView, whose design I wish Django replicated)
* I think it’s interesting to think about if we could get components to re-render based on server sent events somehow, and not only after actions from the user. I think this is a fairly common use-case, that you often encounter in the real world.

@rtr1 About state handling, I think we can leave it to the component to decide what to do: let the user add a inc-method to the component, and then have the users call that to increment and persist that, so that the next reload has the updated value.

**Comment by rtr1** [comment]
@EmilStenstrom there’s an example of using SSE with Django [here](https://valberg.dk/django-sse-postgresql-listen-notify.html).

**Reply by EmilStenstrom** [reply]
Yeah, I think a good tradeoff would be to limit this library to:

1. A django view that sets up a websocket/SSE channel where component updates can be sent to the frontend
2. A javascript that can receive events and re-render a component efficiently
3. Some python code that a component can use to trigger a re-render with new data

If that specific site decides to use Postgres or Redis or just cron-jobs to trigger the re-renders should be up to the specific site.

**Comment by nerdoc** [comment]
A few thoughts to your answers, thanks by the way for taking this up.

IMHO a model centric behavior should be considered, but not enforced. There could be a use case where a component facades a model, and maybe even uses model methods. But there could be some other use cases too. Look here at django-unicorn, or tetraframework. Unicorn takes that really cool, as it enables just "variables" in a component - which in fact are variables in the `view`.

Using django models in all cases is not desireable.

I don't want to reinvent the wheel, because all the other solutions exist. And if you want to use components with all you said in your comments, just use Unicorn. I would definitively use HTMX, because using it helps you

1. to not reinvent a complete Js ecosystem with all the bells and whistles the current solutions struggle (Reactor, Tetra, Unicorn, Hotwire/TurboDjango, Sockpuppet) - it is already there, maintained, and highly stable.
2. You can use HTMX outside of components too. That's a main advantage, because for each small/simple bit of interactivity in all those frameworks, you have to go big and create a full component, with view etc. Sometimes it's just enough to react on a mouseover with a simple POST action, once. Components are overkill then. HTMX solves this.

I understand that you don't want to lock yourself in with HTMX, but like I said, I've tried them all, each of them has their advantages and shortcomings, and I always thought about - why does everyone go and make a "yet another component framework". Why don't they combine the best things together?
I've made a matrix of a few IMHO important things that some framework should  support. I left out Sockpuppet and TurboDjango, as they seem abandoned completely, even if the Discord community is somehow active with Sockpuppet.

| Framework  |Dj-components |Tetra |Reactor |Unicorn | HTMX |
|-------------------|------------------------|-------------|-----------|---|----|
|component based structure|  :heavy_check_mark:   | :heavy_check_mark:   | :heavy_check_mark:  | :heavy_check_mark:  | :heavy_check_mark: | :x: |
|sub-components possible|  :heavy_check_mark:   | :heavy_check_mark:  | :heavy_check_mark:  | :heavy_check_mark: | :heavy_check_mark: (manually) |
|slots   |  :heavy_check_mark:   | :x:  | :x:  | :x: |  :heavy_check_mark: (manually)|
|persistent component server state|  :x:   | :heavy_check_mark:  | :heavy_check_mark:  | :heavy_check_mark: | :heavy_check_mark: (manually) |
|Django model integration| :x: | :heavy_check_mark:   | :heavy_check_mark:   | :heavy_check_mark:   | :heavy_check_mark: (manually) |
|Django forms integration| :x:   | :x:   | :x:   | :x:   | :heavy_check_mark: (manually) |
|Data push support (websockets, SSE)| :x: |  :x:   | :heavy_check_mark: | :x: | :heavy_check_mark: (manually, with hx-ws, hx-sse)|
|well maintained | :heavy_check_mark:   | :skull: (dead?) | minimal |  :heavy_check_mark:   | :heavy_check_mark: |

(I hope I got everything right, please correct me if not.)

In reply to @EmilStenstrom's points:
    > 1. A django view that sets up a websocket/SSE channel where component updates can be sent to the frontend
Second that. Extremely important. A custom channel could be set up for each user. TurboDjango does that very well.

    > 2. A javascript that can receive events and re-render a component efficiently
That is easily (and very eficciently) done using HTMX, using morphdom. works brilliantly, fast, and reliable, using target `#id` s.

    > 3. Some python code that a component can use to trigger a re-render with new data
All frameworks do that automatically after calling some of the compoents' methods. I think Tetraframework does that best. It has a `@public` decorator that marks a method as publicly available in the frontend, which is the most beautiful approach IMHO.
Components must have methods, IMHO they should NOT expose the model's methods, as this could be security relevant. The component is the facade. I don't think it's a good idea to let the compoent "provide" URLs (by manually `get_urls()`), it should create them manually for each of their "public" methods.
HTMX (and maybe a helper templatetag) can call them then easily. HTMX has the flexibility that other frameworks don't have, and you don't have to (and want to) reinvent that wheel again...

1. If Tetra would be maintained actively, and had support for models/forms, and loading templates from a file (and not within a string), I would use Tetra. All the constructive issues at their Github site are by me, and the author never responded again...
2. If Unicorn would be faster (it's relatively slow regarding the component updates, reeeaaallyy slow), and would support forms, I would use it.
3. If Reactor would have a stable, clean API and be more maintained, I would use it.

Maybe a framework agnostic approach would be really cool, and HTMX could be one "plugin" that adds reactability. BUT, that would be even more work.

**IF** you want to use React, Vue, Svelte etc. with django-components, you don't eveen need persistent state and all I talked about, because all you need is an (e.g. REST) API. Going into Server state **without** using a monolithic frontend Js framework is the exact opposite of them. Integrating Django plugins frameworks (like yapsy, or my own [GDAPS](gdaps.readthedocs.io)) need to have components server-side rendered and integratable - this is only possible with things like django-components. And adding reactivity to that would be a killer feature IMHO. There is nothing out there (that I know of) that matches the stability, speed and ease of use of HTMX - and I would definitely not recreate all of their work and reinvent the wheel with a custom framework when everything is already there...

Just my 2 cents.

**Comment by dylanjcastillo** [comment]
I've been playing around a bit with this, and even though, I couldn't wrap my head around the idea of the websockets/SSE channel to update components, I got a simple HTTP-only version of these HTMX enhanced components: 

**Without backend state:**

components/urls.py
```python
from django.urls import path
from components.stateless_counter import StatelessCounterComponent

urlpatterns = [
    path(
        "stateless_counter/<str:method>/<int:id>",
        StatelessCounterComponent.as_view(),
        name="stateless_counter",
    ),
]
```

components/stateless_counter.py
```python
from django_components import component


@component.register("stateless_counter")
class StatelessCounterComponent(component.Component):
    def get_context_data(self, id, **kwargs):
        return {"count": 0, "id": id}

    def post(self, request, method, id, *args, **kwargs):
        if method == "inc":
            count = int(request.POST.get("count", 0)) + 1
        elif method == "dec":
            count = int(request.POST.get("count", 0)) - 1
        context = {
            "count": count,
            "id": id,
        }
        slots = {}
        return self.render_to_response(context, slots)

    template = """
        <div id="stateless-counter-component-{{ id }}" 
             hx-target="#stateless-counter-component-{{ id }}" 
             hx-vals='{ "count": {{ count}} }'>
            <div>Count: {{ count }}</div>
            <button type="button" hx-post="{% url 'stateless_counter' method='dec' id=id %}"> -1 </button>
            <button type="button" hx-post="{% url 'stateless_counter' method='inc' id=id %}"> +1 </button>
        </div>
    """
```

**With backend state:**

components/urls.py
```python
from django.urls import path
from components.counter import CounterComponent

urlpatterns = [
    path("counter/<str:method>/<int:id>", CounterComponent.as_view(), name="counter"),
]
```

counter.py
```python
from django_components import component
from app.models import Counter


@component.register("counter")
class CounterComponent(component.Component):
    model = Counter

    def get_context_data(self, id, **kwargs):
        counter, _ = self.model.objects.get_or_create(id=id) # probably you don't want a create there, but just for testing
        return {"count": counter.count, "id": id}

    def post(self, request, method, id, *args, **kwargs):
        counter, _ = self.model.objects.get_or_create(id=id) # same applies here

        if method == "inc":
            counter.count += 1
            counter.save()
        elif method == "dec":
            counter.count -= 1
            counter.save()

        context = {
            "count": counter.count,
            "id": id,
        }
        slots = {}
        return self.render_to_response(context, slots)

    template = """
        <div id="counter-component-{{ id }}" hx-target="#counter-component-{{ id }}">
            <div>Count: {{ count }}</div>
            <button type="button" hx-post="{% url 'counter' method='dec' id=id %}"> -1 </button>
            <button type="button" hx-post="{% url 'counter' method='inc' id=id %}"> +1 </button>
        </div>
    """
```
https://github.com/EmilStenstrom/django-components/assets/7218756/d8b611e6-cfc7-4e6a-80be-f722908c53c5

**Reply by EmilStenstrom** [reply]
This is exciting! I really think this could be included in the documentation. Is there anything from this code sample that you think could be improved by including some boilerplate in the library?

**Reply by dylanjcastillo** [reply]
I'm still wrapping my head around this.

I think somehow automatically generating an ID would be nice, but not sure how to go about it.

I also have an idea of building a repository of HTMX components using `django-components`. It uses Pyodide so can be hosted as a static site, which would hopefully make it easy for people to contribute to it. 

It's a bit weird, but sort of working for now: https://github.com/iwanalabs/django-htmx-components

I hope to get a decent number of components built by the end of this week and maybe can go back and think about what's missing.

**Reply by EmilStenstrom** [reply]
Making a sharable repository of components is a really good way to understand the things required of django-components. 

Crazy to run django on the client side, but whatever works! :)

I've thought about automatically generating an ID. I think the tricky part is not making that a part of "things you have to know to use django-components". Maybe it could be as easy as modifying examples to have `self.id = uuid.uuid4()` in `__init__()`?

**Reply by dylanjcastillo** [reply]
Oh, you're right. It was as simple as defining `self.id` in `__init__()`. Should we do that in `Component` ourselves? It'd be useful to provide the user with an `id` per component.

I've come up with a pattern that I'm starting to like where you keep all the components related to the same functionality in the same directory and have the URLs there, which you then include in `components/urls.py`: https://github.com/dylanjcastillo/django-htmx-components/tree/main/src/components/bulk_update. It makes it very easy to reason about the component logic.

I also realized that if you include a template such as `_base.html` you need to manually pass global variables such as `csfr_token`. I wonder if this is how it should be.

**Reply by dylanjcastillo** [reply]
I have a few of these now at: https://components.iwanalabs.com/

**Reply by nerdoc** [reply]
Whow, @dylanjcastillo that is exactly what I wanted to achieve. 
1. Is it possible to customize them? I am using [tabler.io](https://tabler.io) as base and would like to use that as base - how easy is it to customize django-htmx-components to another CSS framework?
2. Does it NEED Pyodide?

BTW, your link above to django-htmx-components seems to be broken.

**Reply by EmilStenstrom** [reply]
@nerdoc I'll reply in Dylans place:
1. All the code is right there! The templates have no CSS included at all, so if you want to add stuff you can either subclass them, or copy and change them to use whatever styling framework you want.
2. No, Pyodide is just for demonstration purposes, all the code needed is listed to the right (click the dropdown to see all the files)

**Reply by nerdoc** [reply]
That's great news and exactly what I was looking for. ATM I am using [django-web-components](https://github.com/Xzya/django-web-components) because of its easier syntax, but I think I am going to flip back to django-components again, especially because you are working on #232 and it seems to be a really active project :smile:

**Reply by dylanjcastillo** [reply]
Now there's also a faster non-pyodide version: https://dhc.iwanalabs.com/

**Reply by EmilStenstrom** [reply]
@nerdoc #232 is now merged. Enjoy refactoring your code ;)

**Reply by nerdoc** [reply]
Thanks, I read along all the time. Great work!

**Comment by dylanjcastillo** [comment]
I also built a small POC with SSE using redis:


https://github.com/EmilStenstrom/django-components/assets/7218756/7e99da71-b96e-41fd-aa46-62a9cd6ce9d3


components/notifications.py
```python
import json
from typing import AsyncGenerator
import redis.asyncio as redis
from django.http import StreamingHttpResponse
from django_components import component


@component.register("notification")
class NotificationComponent(component.Component):
    template = """
    <div style="text-color: {{color}};" role="alert">
        <span style="font-weight: bold;">{{ title }}</span> {{ message }} 
    </div>
    """


notification_component = NotificationComponent(
    registered_name="notification",
)


def sse_message(event_id: int, event: str, data: str) -> str:
    data = data.replace("\n", "")
    return f"id: {event_id}\n" f"event: {event}\n" f"data: {data.strip()}\n\n"


r = redis.from_url("redis://localhost")


async def get_component_updates(*args, **kwargs) -> AsyncGenerator[str, None]:
    async with r.pubsub() as pubsub:
        await pubsub.subscribe("notifications_channel")
        try:
            while True:
                message = await pubsub.get_message(ignore_subscribe_messages=True)
                if message is not None:
                    notification_data = json.loads(message["data"].decode())
                    sse_message_rendered = sse_message(
                        notification_data["id"],
                        "notification",
                        notification_component.render(
                            {
                                "title": notification_data["title"],
                                "message": notification_data["message"],
                                "color": notification_data["color"],
                            }
                        ),
                    )
                    yield sse_message_rendered
        finally:
            await r.aclose()


async def stream_component(request):
    return StreamingHttpResponse(
        streaming_content=get_component_updates(),
        content_type="text/event-stream",
    )
```

components/urls.py
```python
from django.urls import path
from components.notifications import stream_component

urlpatterns = [
    path(
        "notification/",
        stream_component,
        name="stream_notification",
    ),
]
```

index.html
```html
{% include "_base.html" %}
{% block content %}
    <div hx-ext="sse"
         sse-connect="{% url 'stream_notification' %}"
         sse-swap="notification"></div>
    <br>
{% endblock %}
```

Not sure if I'm on the right track with this, but was fun haha. If this makes any sense, I think we could probably integrate some of this stuff without a lot of effort.

**Reply by EmilStenstrom** [reply]
Yes, this is exactly what I was thinking. 

1. User loads the page and gets the page server rendered
2. The SSE channel is set up to the server
3. Server starts sending messages to the clients with updated information without the user having to poll for new data

How are you sending messages in your POC? By writing to redis?

**Reply by dylanjcastillo** [reply]
I used this script:

```python
import redis
import json
import random
import time

REDIS_HOST = "localhost"
REDIS_PORT = 6379
REDIS_CHANNEL = "notifications_channel"

r = redis.StrictRedis(host=REDIS_HOST, port=REDIS_PORT, decode_responses=True)


def create_random_notification():
    """Create a random notification message"""
    return {
        "id": random.randint(1, 1000),
        "title": "Notification " + str(random.randint(1, 100)),
        "message": "This is a random message " + str(random.randint(1, 100)),
        "color": random.choice(["#edafb8", "#f7e1d7", "#f9dcc4", "#b0c4b1"]),
        "timestamp": time.ctime(),
    }


def publish_notification():
    """Publish a random notification to the Redis channel"""
    notification = create_random_notification()
    r.publish(REDIS_CHANNEL, json.dumps(notification))
    print(f"Published: {notification}")


if __name__ == "__main__":
    try:
        while True:
            publish_notification()
            time.sleep(3)
    except KeyboardInterrupt:
        print("Stopped notification publisher")
```

It'd be cool if we could provide an easy interface for people to build something like this. I tried getting something like `Component.as_live_view()` to work, but couldn't. If we go for a more functional approach we might need to make some changes/additions to the component registration process.

**Reply by EmilStenstrom** [reply]
This is really inspiring! There's almost nothing that is specific to SSE in the component which is nice. This means that a site built with django-components could be progressively enhanced (in the "JS is optional" sense) to auto-update with very little additional code. Keeping the additional code needed as minimal as possible is the real challenge here I think. And if we could get this API working it would be a really strong advantage of using django-components.

What was the challenges of getting `Component.as_live_view()` working?

Maybe we could have some middleware that wraps each component with an unique ID and the correct hx-attributes (needs to be configurable not to be HTMX specific).

Additionally I like the idea of a single view that you add to your urls.py where each component connects to receive auto-updates.

**Reply by dylanjcastillo** [reply]
Sorry for the slow response. Had a bit too many things on my plate these days.

This is actually possible right now, but I hadn't found a way to make it work before. I found an article that explained the [right way to use async CBVs](https://medium.com/@bruno.fosados/django-async-class-based-views-acbv-5986c4511ae6).

**components/notification.py**
```python
import asyncio
import json
from typing import AsyncGenerator

import redis.asyncio as redis
from django.http import StreamingHttpResponse
from django.utils.decorators import classonlymethod
from django_components import component

r = redis.from_url("redis://localhost")


def sse_message(event_id: int, event: str, data: str) -> str:
    data = data.replace("\n", "")
    return f"id: {event_id}\n" f"event: {event}\n" f"data: {data.strip()}\n\n"


class NotificationComponent(component.Component):

    @classonlymethod
    def as_live_view(cls, **initkwargs):
        view = super().as_view(**initkwargs)
        view._is_coroutine = asyncio.coroutines._is_coroutine
        return view

    template = """
    <div style="color: {{color}};" role="alert">
        <span style="font-weight: bold;">{{ title }}</span> {{ message }} 
    </div>
    """

    async def streaming_response(self, *args, **kwargs) -> AsyncGenerator[str, None]:
        async with r.pubsub() as pubsub:
            await pubsub.subscribe("notifications_channel")
            try:
                while True:
                    message = await pubsub.get_message(
                        ignore_subscribe_messages=True, timeout=1
                    )
                    if message is not None:
                        notification_data = json.loads(message["data"].decode())
                        sse_message_rendered = sse_message(
                            notification_data["id"],
                            "notification",
                            self.render(
                                {
                                    "title": notification_data["title"],
                                    "message": notification_data["message"],
                                    "color": notification_data["color"],
                                }
                            ),
                        )
                        yield sse_message_rendered
                    await asyncio.sleep(0.1)
            finally:
                await r.aclose()

    async def get(self, request, *args, **kwargs):
        return StreamingHttpResponse(
            streaming_content=self.streaming_response(),
            content_type="text/event-stream",
        )
```

**components/urls.py***
```python
from django.urls import path
from components.notification import NotificationComponent

urlpatterns = [
    path(
        "notification/",
        NotificationComponent.as_live_view(),
        name="stream_notification",
    ),
]
```
